moby/moby
# add dockerd-rootless-setuptool

**- What I did**

Added `contrib/dockerd-rootless-setuptool`.

The tool is expected to be used by https://get.docker.com/rootless script.

Can be included in RPM/DPKG packages as well.

Fix https://github.com/moby/moby/issues/40707
Relates to https://github.com/moby/moby/issues/40761

**- How I did it**
Rewrote https://get.docker.com/rootless in Go except the curl stuff.
The curl stuff is expected to be kept in https://get.docker.com/rootless .

Also, the storage driver is not detected in the new tool because the daemon should be able to choose overlay2/fuse-overlayfs/vfs automatically.

**- How to verify it**

```console
$ dockerd-rootless-setuptool --help
A setup tool for dockerd-rootless.sh. Requires dockerd-rootless.sh to be present in $PATH. See https://docs.docker.com/engine/security/rootless/ for the usage.

Usage:
   [command]

Available Commands:
  check       Check prerequisites
  help        Help about any command
  setup       Setup systemd (if available) and show how to run rootless Docker

Flags:
  -h, --help             help for this command
      --ignore-rootful   ignore rootful Docker (/var/run/docker.sock)
      --skip-iptables    ignore mising iptables

Use " [command] --help" for more information about a command.

$ dockerd-rootless-setuptool setup
FATA[0000] Aborting because rootful Docker (/var/run/docker.sock) is running and accessible. Set --ignore-rootful to ignore.

$ dockerd-rootless-setuptool setup --ignore-rootful
INFO[0000] Installing systemd unit file /home/suda/.config/systemd/user/docker.service 
INFO[0000] + systemctl --user daemon-reload             
INFO[0000] + systemctl --user --no-pager --full status docker.service 
● docker.service - Docker Application Container Engine (Rootless)
     Loaded: loaded (/home/suda/.config/systemd/user/docker.service; disabled; vendor preset: enabled)
     Active: inactive (dead)
       Docs: https://docs.docker.com/engine/security/rootless/
...
INFO[0000] Starting docker.service                      
INFO[0000] + systemctl --user start docker.service      
INFO[0000] + systemctl --user --no-pager --full status docker.service 
● docker.service - Docker Application Container Engine (Rootless)
     Loaded: loaded (/home/suda/.config/systemd/user/docker.service; disabled; vendor preset: enabled)
     Active: active (running) since Thu 2020-05-07 22:39:11 JST; 31ms ago
       Docs: https://docs.docker.com/engine/security/rootless/
   Main PID: 32487 (rootlesskit)
     Memory: 1.6M
     CGroup: /user.slice/user-1001.slice/user@1001.service/docker.service
             └─32487 rootlesskit --net=slirp4netns --mtu=65520 --slirp4netns-sandbox=auto --slirp4netns-seccomp=auto --disable-host-loopback --port-driver=builtin --copy-up=/etc --copy-up=/run --propagation=rslave /usr/local/bin/dockerd-rootless.sh

May 07 22:39:11 suda-ws01 dockerd-rootless.sh[32497]: --disable-host-loopback  prohibit connecting to 127.0.0.1:* on the host namespace
May 07 22:39:11 suda-ws01 dockerd-rootless.sh[32487]: + net=slirp4netns
May 07 22:39:11 suda-ws01 dockerd-rootless.sh[32487]: + [ -z ]
May 07 22:39:11 suda-ws01 dockerd-rootless.sh[32487]: + mtu=65520
May 07 22:39:11 suda-ws01 dockerd-rootless.sh[32487]: + [ -z slirp4netns ]
May 07 22:39:11 suda-ws01 dockerd-rootless.sh[32487]: + [ -z 65520 ]
May 07 22:39:11 suda-ws01 dockerd-rootless.sh[32487]: + [ -z ]
May 07 22:39:11 suda-ws01 dockerd-rootless.sh[32487]: + _DOCKERD_ROOTLESS_CHILD=1
May 07 22:39:11 suda-ws01 dockerd-rootless.sh[32487]: + export _DOCKERD_ROOTLESS_CHILD
May 07 22:39:11 suda-ws01 dockerd-rootless.sh[32487]: + exec rootlesskit --net=slirp4netns --mtu=65520 --slirp4netns-sandbox=auto --slirp4netns-seccomp=auto --disable-host-loopback --port-driver=builtin --copy-up=/etc --copy-up=/run --propagation=rslave /usr/local/bin/dockerd-rootless.sh
INFO[0001] + /usr/local/bin/docker version              
Client:
 Version:           20.03.0-dev
 API version:       1.41
 Go version:        go1.13.10
 Git commit:        904e4f4db
 Built:             Wed Apr 29 21:21:44 2020
 OS/Arch:           linux/amd64
 Experimental:      true

Server:
 Engine:
  Version:          dev
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.10
  Git commit:       c3b3aedfa4
  Built:            Wed Apr 29 21:21:03 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.3.0-520-g3f95bc5f
  GitCommit:        3f95bc5f18a7d9dceb1e7502d2fc6120900c5bfb
 runc:
  Version:          1.0.0-rc10+dev
  GitCommit:        d3ac6ba75a5dfcd5c4c00d7b8863fb1ea098be5d
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
INFO[0001] To control docker service run: systemctl --user (start|stop|restart) docker.service 
INFO[0001] Make sure the following environment variables are set (or add them to ~/.bashrc):
export PATH=/usr/local/bin:$PATH
export DOCKER_HOST=unix:///run/user/1001/docker.sock 
```

**- Description for the changelog**

add dockerd-rootless-setuptool

**- A picture of a cute animal (not mandatory but encouraged)**

:penguin:
